File: 11 Linear Classifiers/Perceptron.R
#Set directory: Run this on source instead of Console!!
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

################################################################################
perceptron <- function(X, y, eta=1e-3, n_iter=1e5){
  X <- as.matrix(X); y <- as.numeric(y)
  X_1 <- cbind(1, X)
  n <- nrow(X_1); p <- ncol(X_1)
  omega <- rep(1/p, p)                  # initialize weight vector
  for (it in 0:(n_iter-1)){
    i <- it %% n + 1                    # R index starts from 1
    if (i == 1) new_idx <- sample(1:n, n, replace=FALSE)
    j <- new_idx[i]
    if(y[j] * omega %*% X_1[j,] < 0) {  # misclassified
      omega <- omega + eta * y[j] * X_1[j,]
    } 
  }
  return (list(omega=omega, y_pred=sign(X_1 %*% omega)))
}

################################################################################
df <- read.csv("../Datasets/fin-ratio.csv")
X <- df[-ncol(df)]
y <- df$HSI
y[y==0] <- -1                         # transform the output to {-1, 1}

set.seed(4002)
model <- perceptron(X, y, eta=0.01, n_iter=1e6)
(conf <- table(model$y_pred, y))      # confusion matrix
sum(diag(conf))/length(y)
model$omega

File: 11 Linear Classifiers/Perceptron.py
import numpy as np

def perceptron(X, y, eta=1e-3, n_iter=1e5):
    X, y = np.array(X), np.array(y)
    X_1 = np.c_[np.ones(len(y)), X]
    n, p = X_1.shape
    omega = np.repeat(1/p, p)           # initialize weight vector            
    for it in range(int(n_iter)):
        i = it % n
        if i == 0:                      # Python index starts from 0
            new_idx = np.random.choice(np.arange(n), n, replace=False)
        
        j = new_idx[i]
        if y[j] * omega @ X_1[j] < 0:   # misclassified
            omega += eta * y[j] * X_1[j]
    
    return omega, np.sign(X_1 @ omega)

#%%
import pandas as pd
from sklearn.metrics import confusion_matrix

df = pd.read_csv("../Datasets/fin-ratio.csv")
X, y = df.drop(columns=["HSI"]), df["HSI"]
y = y.replace(0, -1)				    # transform the output to {-1, 1}

np.random.seed(4002)
omega, y_pred = perceptron(X, y, eta=0.01, n_iter=1e6)
conf = confusion_matrix(y_pred, y) 	    # confusion matrix
print(conf)
print(sum(np.diag(conf))/len(y))
print(omega)